### ISSUE TITLE
LoRA Fine-tuning Results in Visual Artifacts Despite Loss Convergence - Seeking Advice

### ISSUE THREAD

**Issue by daedaluser** [Tongyi-MAI/Z-Image#138]
Hello! Thank you for open-sourcing the base model.
I'm encountering some issues while performing LoRA fine-tuning on the base model using 400 anime images (approximately 80 images per single character) with the goal of learning art style and facial features. I've labeled all images with captions.
Training Configuration:
I fine-tuned the following modules: to_v, to_k, w1, w2, w3
Key hyperparameters:
lora_rank = 32
lora_alpha = 64
lora_dropout = 0.0
Learning rate: 1e-4
Optimizer: AdamW with betas=(0.9, 0.999), weight_decay=0.01
LR Scheduler: CosineAnnealingWarmRestarts(optimizer, T_0=max(1, total_steps//10), T_mult=2, eta_min=1e-6)
Issue Description:
The loss converged from 10+ down to around 5, but the generated images exhibit visual artifacts/corrupted pixels (坏点).
Could you please provide some valuable suggestions regarding LoRA fine-tuning? Specifically, I'm wondering if there are recommended adjustments for the rank/alpha ratio, learning rate, or target modules when simultaneously learning style and character features.
Thank you in advance for your guidance!

**Comment by daedaluser** [comment]
@ThanhNguyxn 
Thank you for the quick response and clarification!
I am using a custom training script based on diffusers with peft for LoRA, not DiffSynth-Studio. Below are the complete configuration details and the minimal reproducible code snippets:

**Training Details:**
Base Model: Z-Image (Flow Matching 8-step diffusion model)
Resolution: 896×896
Dataset: 400 anime images (single character, ~80 images/character), JSON metadata with BLIP captions
Batch size: 1 (Gradient Accumulation: 2 steps → Effective batch size: 2)
Total Steps: ~4,000 steps (20 epochs)
Loss Convergence: Started ~10.5, converged to ~4.5-5.0

**LoRA Configuration:**
```
target_modules = ["to_q", "to_k", "to_v", "to_out.0"] # Target modules (Attention layers only, excluding FFN)

lora_config = LoraConfig(
    r=32,
    lora_alpha=64,
    target_modules=target_modules,
    lora_dropout=0.0,
    bias="none",
    init_lora_weights="gaussian",
)

optimizer = AdamW(..., lr=2e-5, betas=(0.9, 0.999), weight_decay=0.01)# Optimizer & Scheduler
lr_scheduler = CosineAnnealingWarmRestarts(optimizer, T_0=400, T_mult=2, eta_min=1e-6)

```
**VAE Encoding Logic (Potential Issue Area):**
I noticed the model uses custom VAE scaling factors. Here is my encoding implementation:

```
def encode_vae(self, images):
    with torch.no_grad():
        h = self.vae.encoder(images)
        mean, logvar = torch.chunk(h, 2, dim=1)
        std = torch.exp(0.5 * logvar)
        latents = mean + std * torch.randn_like(std)
        latents = (latents / self.vae_scaling_factor) + self.vae_shift_factor     # Custom scaling matching inference logic
        return latents.to(dtype=self.dtype)        # vae_scaling_factor = 0.3611, vae_shift_factor = 0.1159
```

**Loss Computation (Flow Matching):**
```
def add_noise(self, latents, noise, t):
    t = t.view(-1, 1, 1, 1)
    return (1 - t) * latents + t * noise  # Linear interpolation

velocity_target = noise - latents# Velocity target: v = noise - x_0
loss = F.mse_loss(velocity_pred.float(), velocity_target.float())

```
**Artifact Description ("坏点"):**
The artifacts appear as random bright/dark pixel spots (salt-and-pepper noise pattern), primarily in:
Flat color regions (skies, walls)
Character skin areas
Occasional grid-like patterns at latent block boundaries

**Key Questions:**
Target Modules: Is targeting to_q, to_k, to_v, to_out.0 sufficient, or should I include w1, w2, w3 (FFN layers) for style/character LoRA?
 I noticed the base model has all_patch_size=(2,) - does this affect which layers should be trained?
VAE Scaling: My script handles vae_scaling_factor (0.3611) and shift_factor (0.1159) manually. Is this correct for Z-Image？
LR Scheduler: Could CosineAnnealingWarmRestarts with T_mult=2 cause instability? Would constant LR or cosine decay without restarts be safer for 4k steps?
Rank/Alpha Ratio: Is alpha=64 with rank=32 (2:1 ratio) too aggressive for only 400 images? Should I reduce to alpha=32 or increase dropout?

**Additional Context:**
Training in bfloat16 with torch.cuda.amp.autocast
Using unconditional_prob=0.1 for CFG training
Resume from checkpoint was used (loaded previous LoRA weights as initialization)
Thank you for your guidance!

**Comment by ThanhNguyxn** [comment]
Thank you for the detailed breakdown! This is extremely helpful. Since this repo doesn't include official LoRA training scripts, I can provide targeted guidance on potential issues:

**VAE Scaling & Latent Space:**
The VAE scaling factors (0.3611, 0.1159) you're using—where did you derive these? For Z-Image's VAE, verify these match the inference pipeline exactly. Artifacts in flat regions often indicate latent scaling mismatch (causes quantization-like errors or loss of precision).

**Target Modules:**
Targeting attention layers (to_q, to_k, to_v, to_out.0) is reasonable for style/character LoRA. Including FFN (w1, w2, w3) typically requires more data and can increase overfitting with only 400 images. **Recommendation:** Try attention-only first; if results improve but remain limited, FFN layers may help, but expect artifacts to potentially worsen without proper regularization.

**Loss Convergence ≠ Visual Quality:**
A well-converged loss (10.5→4.5-5.0) doesn't guarantee artifact-free outputs. The velocity prediction (v = noise - x₀) is mathematically correct, but salt-and-pepper noise in flat regions suggests:
- **Latent distribution shift** from VAE scaling miscalibration
- **Overfitting on small dataset** (400 images) despite high α=64
- **Numerical stability** in low-rank projections with high α

**Concrete Diagnostics:**
1. **Alpha/Rank Ratio:** Try reducing α from 64→32 (keeping r=32, so 1:1 ratio) for 400 images
2. **LR Scheduler:** CosineAnnealingWarmRestarts with T_mult=2 can cause periodic spikes. Simpler option: LinearLR warmup + CosineAnnealingLR without restarts
3. **Dropout:** Set lora_dropout=0.1-0.15 for regularization with small dataset
4. **Test VAE:** Generate a latent sample, encode/decode without LoRA to isolate VAE scaling issues

**Next Steps:**
If the issue persists, please share (or link):
- Exact VAE scaling factor derivation
- Sample output images showing artifacts
- Loss curve plot (training vs. val)

Consider posting this to DiffSynth-Studio issues or diffusers discussions—they have more direct experience with LoRA training on diffusion models. They may recognize these artifacts immediately.
